internal/handler/msg_saslstart.go

```diff
@@ -129,3 +147,117 @@ func saslStartPlain(doc *types.Document) (string, string, error) {
+	adminDB, err := h.b.Database("admin")
+	if err != nil {
+		return scram.StoredCredentials{}, err
+	}
+
+	usersCol, err := adminDB.Collection("system.users")
+	if err != nil {
+		return scram.StoredCredentials{}, err
+	}
+
+	var filter *types.Document
+	filter, err = usersInfoFilter(false, false, "", []usersInfoPair{
+		{username: username, db: dbName},
+	})
+
+	if err != nil {
+		return scram.StoredCredentials{}, err
+	}
+
+	// Filter isn't being passed to the query as we are filtering after retrieving all data
+	// from the database due to limitations of the internal/backends filters.
+	qr, err := usersCol.Query(ctx, nil)
+	if err != nil {
+		return scram.StoredCredentials{}, err
+	}
+
+	defer qr.Iter.Close()
+
+	for {
+		_, v, err := qr.Iter.Next()
+
+		if errors.Is(err, iterator.ErrIteratorDone) {
+			break
+		}
+
+		if err != nil {
+			return scram.StoredCredentials{}, lazyerrors.Error(err)
+		}
+
+		matches, err := common.FilterDocument(v, filter)
+		if err != nil {
+			return scram.StoredCredentials{}, lazyerrors.Error(err)
+		}
+
+		if matches {
+			if !v.Has("credentials") {
+				return scram.StoredCredentials{}, handlererrors.NewCommandErrorMsgWithArgument(
+					handlererrors.ErrMechanismUnavailable,
+					"User has no authentication credentials registered",
+					"credentials",
+				)
+			}
+
+			credentials := must.NotFail(v.Get("credentials")).(*types.Document)
+
+			if !credentials.Has("SCRAM-SHA-256") {
+				return scram.StoredCredentials{}, handlererrors.NewCommandErrorMsgWithArgument(
+					handlererrors.ErrMechanismUnavailable,
+					"User has no SCRAM-SHA-256 based authentication credentials registered",
+					"SCRAM-SHA-256",
+				)
+			}
+
+			cred := must.NotFail(credentials.Get("SCRAM-SHA-256")).(*types.Document)
+
+			salt := must.NotFail(base64.StdEncoding.DecodeString(must.NotFail(cred.Get("salt")).(string)))
+			storedKey := must.NotFail(base64.StdEncoding.DecodeString(must.NotFail(cred.Get("storedKey")).(string)))
+			serverKey := must.NotFail(base64.StdEncoding.DecodeString(must.NotFail(cred.Get("serverKey")).(string)))
+
+			return scram.StoredCredentials{
+				KeyFactors: scram.KeyFactors{
+					Salt:  string(salt),
+					Iters: int(must.NotFail(cred.Get("iterationCount")).(int32)),
+				},
+				StoredKey: storedKey,
+				ServerKey: serverKey,
+			}, nil
+		}
+	}
+
+	return scram.StoredCredentials{}, errors.New("user not found")
+}
+
+func (h *Handler) saslStartSCRAMSHA256(ctx context.Context, doc *types.Document) (string, error) {
+	var payload []byte
+
+	binaryPayload, err := common.GetRequiredParam[types.Binary](doc, "payload")
+	if err == nil {
+		payload = binaryPayload.B
+	}
+
+	dbName, err := common.GetRequiredParam[string](doc, "$db")
+	if err != nil {
+		return "", err
+	}
+
+	scramServer, err := scram.SHA256.NewServer(func(username string) (scram.StoredCredentials, error) {
+		return h.credentialLookup(ctx, username, dbName)
+	})
+	must.NoError(err)
```

Comment:
Should we handle error gracefully rather than panic?
```suggestion
	if err != nil{
		return "", err
	}
```